internal/server/text_syncronization.go

```diff
@@ -0,0 +1,362 @@
+	lineOffset := lineStarts[lineIndex]
+
+	// Extract the content of the requested line
+	lineEndOffset := len(content)
+	if lineIndex+1 < len(lineStarts) {
+		lineEndOffset = lineStarts[lineIndex+1] - 1 // -1 to exclude the newline character
+	}
+
+	// Ensure we don't go beyond the end of content
+	if lineOffset >= len(content) {
+		return len(content)
+	}
+
+	lineContent := content[lineOffset:min(lineEndOffset, len(content))]
+
+	// Convert UTF-16 character offset to UTF-8 byte offset
+	utf8Offset := utf16OffsetToUTF8(string(lineContent), int(position.Character))
+
+	// Ensure the final offset doesn't exceed the content length
+	return lineOffset + utf8Offset
+}
+
+var mu sync.Mutex
+
+// getDiagnostics generates diagnostic information for a specific file.
+// It performs two checks:
+// 1. AST parsing - reports syntax errors
+// 2. Type checking - reports type errors
+//
+// If AST parsing fails, only syntax errors are returned as diagnostics.
+// If AST parsing succeeds but type checking fails, type errors are returned.
+// Returns a slice of diagnostics and an error (if diagnostic generation failed).
+func (s *Server) getDiagnostics(path string) ([]Diagnostic, error) {
+	var diagnostics []Diagnostic
+	mu.Lock()
+	defer mu.Unlock()
+
+	proj := s.getProj()
+
+	// 1. Get AST diagnostics
+	// Parse the file and check for syntax errors
+	astFile, err := proj.AST(path)
+	if err != nil {
+		var (
+			errorList gopscanner.ErrorList
+			codeError *gogen.CodeError
+		)
+		if errors.As(err, &errorList) {
+			// Handle parse errors.
+			for _, e := range errorList {
+				diagnostics = append(diagnostics, Diagnostic{
+					Severity: SeverityError,
+					Range:    s.rangeForASTFilePosition(astFile, e.Pos),
+					Message:  e.Msg,
+				})
+			}
+		} else if errors.As(err, &codeError) {
+			// Handle code generation errors.
+			diagnostics = append(diagnostics, Diagnostic{
+				Severity: SeverityError,
+				Range:    s.rangeForPos(codeError.Pos),
+				Message:  codeError.Error(),
+			})
+		} else {
+			// Handle unknown errors (including recovered panics).
+			diagnostics = append(diagnostics, Diagnostic{
+				Severity: SeverityError,
+				Message:  fmt.Sprintf("failed to parse spx file: %v", err),
+			})
+		}
+	}
+
+	if astFile == nil {
+		return diagnostics, nil
+	}
+
+	handleErr := func(err error) {
+		if typeErr, ok := err.(types.Error); ok {
+			diagnostics = append(diagnostics, Diagnostic{
+				Severity: SeverityError,
+				Range:    s.rangeForPos(typeErr.Pos),
+				Message:  typeErr.Msg,
+			})
+		}
+	}
+
+	// 2. Get type checking diagnostics
+	// Perform type checking on the file
+	_, _, err, _ = proj.TypeInfo()
+	if err != nil {
+		// Add type checking errors to diagnostics
+		switch err := err.(type) {
+		case errors.List:
+			for _, e := range err {
+				handleErr(e)
+			}
+		default:
+			handleErr(err)
+		}
+	}
+
+	return diagnostics, nil
+}
+
+// fromPosition converts a token.Position to an LSP Position.
```

Comment:
I'd suggest moving `fromPosition` and the methods below to the `server.go` file since they are too common.